packages/coinstac-ui/app/main/index.js

```diff
@@ -198,6 +198,25 @@ loadConfig()
       mainWindow.webContents.send('prepare-consortia-files', res);
     });
 
+    /**
+     * IPC Listener to generate provenance
+     */
+    ipcMain.on('generate-provenance', async (event, { userId, appDirectory, runs }) => {
+      runs.forEach((run) => {
+        const runProvenanceDirectory = path.join(appDirectory, 'output', userId, run.id);
+
+        if (!fs.existsSync(runProvenanceDirectory)) {
+          fs.mkdirSync(runProvenanceDirectory, { recursive: true });
+        }
+
+        const provenanceFilePath = path.join(runProvenanceDirectory, 'provenance.json');
+
+        if (!fs.existsSync(provenanceFilePath)) {
+          fs.writeFileSync(provenanceFilePath, JSON.stringify(run, null, 2));
+        }
+      });
+    });
+
```

Comment:
This is an async function, why use sync fs methods? You're tying down the main thread doing so